src/zlib_stubs.c

```diff
@@ -294,8 +299,7 @@ CAMLprim value zlib_set_header(value vstrm, value vheader)
     /* this string is _not_ expected to be zero-terminated */
     len = caml_string_length(Field(Field(vheader,4),0));
     header->extra_len = len;
-    header->extra = caml_stat_alloc(len);
-    memcpy(header->extra, String_val(Field(Field(vheader,4),0)), len);
+    memcpy(wrap->header->extra, String_val(Field(Field(vheader,4),0)), len);
```

Comment:
This whole thing should raise if the buffers are too small:
```
if (caml_string_length() + 1 >= header->extra_len)
   caml_invalid_argument("too long extra string");
```

But maybe keep allocating custom chunks here and fix the finalization procedure.